### ISSUE TITLE
Altering `string` to `text` fails in Oracle

### ISSUE THREAD

**Issue by rijkvanzanten** [knex/knex#4462]
# Environment

Knex version: `knex@0.95.4`
Database + version: Oracle 18XE
OS: Docker

# Bug / Feature Request

When changing the type of a column from "string" to "text" in Knex, it works across databases except for Oracle. In the following case:

```js
await knex.schema.createTable('example_table', (table) => {
	table.string('example_field');
});

await knex.schema.alterTable('example_table', (table) => {
	table.text('example_field').alter();
});
```

Oracle throws the following error:

```
alter table "example_table" modify "example_field" clob
ORA-22858: invalid alteration of datatype
```

because it's not allowed/supported to directly change a `varchar` to a `clob` in OracleDB.

---

Now for the difficult question: 

Do we consider this a "You're holding it wrong" issue, as this is simply not supported in OracleDB, or do we consider this a feature request in Knex to abstract this away? 

To me, it's more the latter, as Knex already includes a bunch of abstractions to make alterations work in SQLite for example, so to me it makes the most sense to smoothen out these differences across database vendors, but I'd love to hear your thoughts @kibertoad, @elhigu. Cc @nickrum, @aidenfoxx.

This can be abstracted away in a very similar fashion to SQLite, where we first create a new column with a temporary name, copy over the data, remove the old column, and rename the new one to the old name.

**Comment by aidenfoxx** [comment]
I tend to agree. If the point of using knex is to aggregate various SQL syntax's into a common interface, then it should stand to reason that the common interface should always behave in the same way (as much as possible). And this seems like an instance where knex could handle the odd Oracle quirk.

**Comment by elhigu** [comment]
I don't think knex should try to emulate features of different databases under a single API unless those features are next to trivial to implement. 

I have been thinking already years that the migration/schema building engine should be out of scope of knex query builder and have their own project(s). Schema builder already breaks one of the knex's basic design principles by using multiple queries to emulate things which opens a door to be able to emulate pretty much anything.

As a separate project it wouldn't matter so much if the migration parts evolves to more complicated monster that knex query building  core parts (we dont have  enough maintainers to even take care of the core parts properly) .  Also it would allow easily to select which migration engine to use with knex.

**Comment by mercmobily** [comment]
I am copying [this text](https://github.com/knex/knex/issues/4462#issuecomment-870417329) into #6301

**Comment by mercmobily** [comment]
I accidentally closed it. 


Oracle does not support altering a `varchar2` column to `clob` directly, so the naive `alter table ... modify ... clob` fails with ORA‑22858. If we want to smooth this out like we do for SQLite, the Oracle dialect might need to emulate the change by creating a new column, copying data, dropping the old column and renaming the new one within a sequence that works safely for that engine.

To support this transparently, we would have to create an Oracle‑specific alter flow akin to SQLite’s DDL rewriter (see `lib/dialects/sqlite3/schema/ddl.js:1`), but tailored to Oracle: add temp column, copy values, drop old column, rename temp, and recreate dependent indexes/constraints. This is non‑trivial and would require careful handling of defaults and constraints.

@elhigu's [comment](https://github.com/knex/knex/issues/4462#issuecomment-870417329) does say "careful" and I responded *today* to a help request that dealt with exactly this in sqllite.

Leaving it open for discussion, but...
